### PR TITLE
Fix phpdoc and prefer null as default value

### PR DESCRIPTION
**Pull request by VincentLanglet**
Hi @nicolaasuni , since the PR: https://github.com/tecnickcom/TCPDF/pull/345 by @williamdes 

There was a lot of phpdoc like this:
```
/**
 * @param float $x
 * @param float $y
 **/
public function foo($x = '', $y = '')
```

So if you need to pass the `$y` value, you use `foo('', 42)`.
When using some static analysis tool, the `''` is reported as incorrect since the phpdoc expect a float.

Instead of changing the phpdoc to `float|string`, I preferred to change to `float|null` to not allow every string values.
And in order to be BC, I changed the check `=== ''` to use `TCPDF_STATIC::empty_string` method, which also check for null values.

I also found some others errors in the phpdoc, so I fixed them.

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Friendly ping @nicolaasuni if you have time to take a look.
The current phpdoc have a lot of errors which causes a lot of false positive when using static analysis tool.
Fixing this would be great :)

File: tcpdf.php [hunk 1]
```diff
@@ -15266,11 +15266,11 @@ public function getBarcode() {
 	 * Print a Linear Barcode.
 	 * @param string $code code to print
 	 * @param string $type type of barcode (see tcpdf_barcodes_1d.php for supported formats).
-	 * @param int $x x position in user units (empty string = current x position)
-	 * @param int $y y position in user units (empty string = current y position)
-	 * @param int $w width in user units (empty string = remaining page width)
-	 * @param int $h height in user units (empty string = remaining page height)
-	 * @param float $xres width of the smallest bar in user units (empty string = default value = 0.4mm)
+	 * @param int|null $x x position in user units (null = current x position)
```
Comment (mvorisek) [inline tcpdf.php:15269]: `float|null`

`$x`, `$w` accepts `float` in general

Comment (VincentLanglet) [reply tcpdf.php:15269]: Not sure it's worth reviewing since I got no answer from @nicolaasuni so far :s

Comment (mvorisek) [reply tcpdf.php:15269]: It is, he only do the merges a few times per year :D :(

Comment (VincentLanglet) [reply tcpdf.php:15269]: Then feel free to list me the int|null to change to float|null, I'll do the update ;)

Comment (mvorisek) [reply tcpdf.php:15269]: basically any location/dimension parameter

File: tcpdf.php [hunk 2]
```diff
@@ -11118,17 +11118,17 @@ public function ScaleXY($s, $x='', $y='') {
 	 * Vertical and horizontal non-proportional Scaling.
 	 * @param float $s_x scaling factor for width as percent. 0 is not allowed.
 	 * @param float $s_y scaling factor for height as percent. 0 is not allowed.
-	 * @param int $x abscissa of the scaling center. Default is current x position
-	 * @param int $y ordinate of the scaling center. Default is current y position
+	 * @param int|null $x abscissa of the scaling center. Default is current x position
+	 * @param int|null $y ordinate of the scaling center. Default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11122]:
```suggestion
	 * @param float|null $x abscissa of the scaling center. Default is current x position
	 * @param float|null $y ordinate of the scaling center. Default is current y position
```

File: tcpdf.php [hunk 3]
```diff
@@ -11152,48 +11152,48 @@ public function Scale($s_x, $s_y, $x='', $y='') {
 
 	/**
 	 * Horizontal Mirroring.
-	 * @param int $x abscissa of the point. Default is current x position
+	 * @param int|null $x abscissa of the point. Default is current x position
```
Comment (VincentLanglet) [inline tcpdf.php:11155]:
```suggestion
	 * @param float|null $x abscissa of the point. Default is current x position
```

File: tcpdf.php [hunk 4]
```diff
@@ -11152,48 +11152,48 @@ public function Scale($s_x, $s_y, $x='', $y='') {
...
 	 */
-	public function MirrorH($x='') {
+	public function MirrorH($x=null) {
 		$this->Scale(-100, 100, $x);
 	}
 
 	/**
 	 * Verical Mirroring.
-	 * @param int $y ordinate of the point. Default is current y position
+	 * @param int|null $y ordinate of the point. Default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11166]:
```suggestion
	 * @param float|null $y ordinate of the point. Default is current y position
```

File: tcpdf.php [hunk 5]
```diff
@@ -11152,48 +11152,48 @@ public function Scale($s_x, $s_y, $x='', $y='') {
...
+	public function MirrorV($y=null) {
+		$this->Scale(100, -100, null, $y);
 	}
 
 	/**
 	 * Point reflection mirroring.
-	 * @param int $x abscissa of the point. Default is current x position
-	 * @param int $y ordinate of the point. Default is current y position
+	 * @param int|null $x abscissa of the point. Default is current x position
+	 * @param int|null $y ordinate of the point. Default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11178]:
```suggestion
	 * @param float|null $x abscissa of the point. Default is current x position
	 * @param float|null $y ordinate of the point. Default is current y position
```

File: tcpdf.php [hunk 6]
```diff
@@ -11152,48 +11152,48 @@ public function Scale($s_x, $s_y, $x='', $y='') {
...
 		$this->Scale(-100, -100, $x, $y);
 	}
 
 	/**
 	 * Reflection against a straight line through point (x, y) with the gradient angle (angle).
 	 * @param float $angle gradient angle of the straight line. Default is 0 (horizontal line).
-	 * @param int $x abscissa of the point. Default is current x position
-	 * @param int $y ordinate of the point. Default is current y position
+	 * @param int|null $x abscissa of the point. Default is current x position
+	 * @param int|null $y ordinate of the point. Default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11191]:
```suggestion
	 * @param float|null $x abscissa of the point. Default is current x position
	 * @param float|null $y ordinate of the point. Default is current y position
```

File: tcpdf.php [hunk 7]
```diff
@@ -11244,17 +11244,17 @@ public function Translate($t_x, $t_y) {
 	/**
 	 * Rotate object.
 	 * @param float $angle angle in degrees for counter-clockwise rotation
-	 * @param int $x abscissa of the rotation center. Default is current x position
-	 * @param int $y ordinate of the rotation center. Default is current y position
+	 * @param int|null $x abscissa of the rotation center. Default is current x position
+	 * @param int|null $y ordinate of the rotation center. Default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11248]:
```suggestion
	 * @param float|null $x abscissa of the rotation center. Default is current x position
	 * @param float|null $y ordinate of the rotation center. Default is current y position
```

File: tcpdf.php [hunk 8]
```diff
@@ -11274,44 +11274,44 @@ public function Rotate($angle, $x='', $y='') {
 	/**
 	 * Skew horizontally.
 	 * @param float $angle_x angle in degrees between -90 (skew to the left) and 90 (skew to the right)
-	 * @param int $x abscissa of the skewing center. default is current x position
-	 * @param int $y ordinate of the skewing center. default is current y position
+	 * @param int|null $x abscissa of the skewing center. default is current x position
+	 * @param int|null $y ordinate of the skewing center. default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11278]:
```suggestion
	 * @param float|null $x abscissa of the skewing center. default is current x position
	 * @param float|null $y ordinate of the skewing center. default is current y position
```

File: tcpdf.php [hunk 9]
```diff
@@ -11274,44 +11274,44 @@ public function Rotate($angle, $x='', $y='') {
...
 		$this->Skew($angle_x, 0, $x, $y);
 	}
 
 	/**
 	 * Skew vertically.
 	 * @param float $angle_y angle in degrees between -90 (skew to the bottom) and 90 (skew to the top)
-	 * @param int $x abscissa of the skewing center. default is current x position
-	 * @param int $y ordinate of the skewing center. default is current y position
+	 * @param int|null $x abscissa of the skewing center. default is current x position
+	 * @param int|null $y ordinate of the skewing center. default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11291]:
```suggestion
	 * @param float|null $x abscissa of the skewing center. default is current x position
	 * @param float|null $y ordinate of the skewing center. default is current y position
```

File: tcpdf.php [hunk 10]
```diff
@@ -11274,44 +11274,44 @@ public function Rotate($angle, $x='', $y='') {
...
 	}
 
 	/**
 	 * Skew.
 	 * @param float $angle_x angle in degrees between -90 (skew to the left) and 90 (skew to the right)
 	 * @param float $angle_y angle in degrees between -90 (skew to the bottom) and 90 (skew to the top)
-	 * @param int $x abscissa of the skewing center. default is current x position
-	 * @param int $y ordinate of the skewing center. default is current y position
+	 * @param int|null $x abscissa of the skewing center. default is current x position
+	 * @param int|null $y ordinate of the skewing center. default is current y position
```
Comment (VincentLanglet) [inline tcpdf.php:11305]:
```suggestion
	 * @param float|null $x abscissa of the skewing center. default is current x position
	 * @param float|null $y ordinate of the skewing center. default is current y position
```

File: tcpdf.php (hunk 1 above)
Comment (VincentLanglet) [reply tcpdf.php:15269]:
```suggestion
	 * @param float|null $x x position in user units (null = current x position)
```

File: tcpdf.php [hunk 11]
```diff
@@ -15266,11 +15266,11 @@ public function getBarcode() {
...
 	 * @param string $type type of barcode (see tcpdf_barcodes_1d.php for supported formats).
-	 * @param int $x x position in user units (empty string = current x position)
-	 * @param int $y y position in user units (empty string = current y position)
-	 * @param int $w width in user units (empty string = remaining page width)
-	 * @param int $h height in user units (empty string = remaining page height)
-	 * @param float $xres width of the smallest bar in user units (empty string = default value = 0.4mm)
+	 * @param int|null $x x position in user units (null = current x position)
+	 * @param int|null $y y position in user units (null = current y position)
+	 * @param int|null $w width in user units (null = remaining page width)
+	 * @param int|null $h height in user units (null = remaining page height)
```
Comment (VincentLanglet) [inline tcpdf.php:15272]:
```suggestion
	 * @param float|null $y y position in user units (null = current y position)
	 * @param float|null $w width in user units (null = remaining page width)
	 * @param float|null $h height in user units (null = remaining page height)
```

File: tcpdf.php [hunk 12]
```diff
@@ -15590,10 +15590,10 @@ public function write1DBarcode($code, $type, $x='', $y='', $w='', $h='', $xres='
...
 	 * @param string $code code to print
 	 * @param string $type type of barcode (see tcpdf_barcodes_2d.php for supported formats).
-	 * @param int $x x position in user units
-	 * @param int $y y position in user units
-	 * @param int $w width in user units
-	 * @param int $h height in user units
+	 * @param int|null $x x position in user units
+	 * @param int|null $y y position in user units
+	 * @param int|null $w width in user units
+	 * @param int|null $h height in user units
```
Comment (VincentLanglet) [inline tcpdf.php:15596]:
```suggestion
	 * @param float|null $x x position in user units
	 * @param float|null $y y position in user units
	 * @param float|null $w width in user units
	 * @param float|null $h height in user units
```

File: tcpdf.php [hunk 13]
```diff
@@ -22483,8 +22483,8 @@ public function endTemplate() {
 	 * An XObject Template is a PDF block that is a self-contained description of any sequence of graphics objects (including path objects, text objects, and sampled images).
 	 * An XObject Template may be painted multiple times, either on several pages or at several locations on the same page and produces the same results each time, subject only to the graphics state at the time it is invoked.
 	 * @param string $id The ID of XObject Template to print.
-	 * @param int $x X position in user units (empty string = current x position)
-	 * @param int $y Y position in user units (empty string = current y position)
+	 * @param int|null $x X position in user units (empty string = current x position)
+	 * @param int|null $y Y position in user units (empty string = current y position)
```
Comment (VincentLanglet) [inline tcpdf.php:22487]:
```suggestion
	 * @param float|null $x X position in user units (empty string = current x position)
	 * @param float|null $y Y position in user units (empty string = current y position)
```

File: tcpdf.php [hunk 14]
```diff
@@ -22483,8 +22483,8 @@ public function endTemplate() {
...
 	 * @param int $w Width in user units (zero = remaining page width)
 	 * @param int $h Height in user units (zero = remaining page height)
```
Comment (VincentLanglet) [inline tcpdf.php:22489]:
```suggestion
	 * @param float $w Width in user units (zero = remaining page width)
	 * @param float $h Height in user units (zero = remaining page height)
```

File: tcpdf.php (hunk 1 above)
Comment (VincentLanglet) [reply tcpdf.php:15269]: > basically any location/dimension parameter

I just listed them all then. Can you confirm before I commit them ?

Comment (mvorisek) [reply tcpdf.php:15269]: LGTM 👍

**Review by nicolaasuni — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @nicolaasuni ; would it be possible to release this phpdoc fix ? This would fix a lot of static analyser errors :)

**Pull request merged**